app/services/membership/withdraw_service.rb

```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class Membership::WithdrawService
+  attr_reader :group, :user
+
+  class MembershipMustHaveDefaultPost < StandardError; end
+
+  def initialize(group, user)
+    @group = group
+    @user = user
+  end
+
+  def self.call(group, user)
+    new(group, user).perform
+  end
+
+  def perform
+    membership = user.membership_for(group)
+    raise MembershipMustHaveDefaultPost unless membership.has_post?(PostType::DEFAULT_POST_ID)
+
+    membership.destroy!
```

Comment:
When somebody applies to a group the leader receives a notification. Have you checked what happens with it on destroy? If I remember well the notification system handles the removal of the target, but it leaves an 'empty' notification. If yes, it'd be a nicer approach to remove it also.